DOCSP-42346: Upgrade driver (#91)
mongodb/docs-scala: source/upgrade.txt

- When using the ``MONGODB-OIDC`` authentication mechanism, you must
  not include comma characters in the ``authMechanismProperties`` connection
  string value.

.. _scala-breaking-changes-v5.0:

Version 5.0 Breaking Changes
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This driver version introduces the following breaking changes:

- Introduces the following changes to the ``ConnectionId`` class:

  - The ``ConnectionId`` constructor now accepts a value of type ``long`` as its second
    parameter instead of type ``int``. Similarly, the constructor now accepts a value of
    type ``Long`` as its third parameter instead of type ``Integer``. Because this change breaks
    binary compatibility, recompile any existing code that calls the
    ``ConnectionId`` constructor.

  - The ``withServerValue()`` method now accepts a parameter of type ``long`` rather than
    type ``int``. This change breaks binary compatibility, so you must recompile any code
    that calls the ``withServerValue()`` method.

  - The ``getServerValue()`` method now returns a value of type ``Long`` instead of type
    ``Integer``. Similarly, the ``getLocalValue()`` method returns a value of type

>     ``long`` instead of type ``int``. Because this change breaks both binary and source

Author: That's how they're formatted in the source code, lowercase are primitive data types and uppercase are class types - https://github.com/mongodb/mongo-java-driver/blob/49cecd9cafc70cb56467f310a525d77a4d45eded/driver-core/src/main/com/mongodb/connection/ConnectionId.java

  change breaks binary compatibility and requires recompiling, but does not
  require code changes.

- Removes the ``Filters.eqFull()`` method, released
  exclusively in ``Beta``, which allowed you
  to construct an equality filter when performing a vector search.
  You can use the ``Filters.eq()`` method when instantiating a
  ``VectorSearchOptions`` type, as shown in the following code:

  .. code-block:: java

     VectorSearchOptions opts = vectorSearchOptions().filter(eq("x", 8));

.. _scala-breaking-changes-v5.0-observables:

- Removes the
  ``org.mongodb.scala.ObservableImplicits.ToSingleObservableVoid`` implicit
  class. This means the ``org.reactivestreams.Publisher[Void]`` type no longer
  converts automatically to ``org.mongodb.scala.SingleObservable[Void]``. The
  API also exposes ``org.mongodb.scala.Observable[Unit]`` instead of
  ``org.mongodb.scala.Observable[Void]``.

  For more information, see the `Observable trait in the Scala API documentation
  <https://mongodb.github.io/mongo-java-driver/5.0/apidocs/mongo-scala-driver/org/mongodb/scala/Observable.html>`__.

- Changes how ``ClusterSettings`` computes

>   ``ClusterConnectionMode``, making it more consistent by using the specified

Author: changed to "setting"

  - ``ConnectionCheckOutFailedEvent(ServerId, long, Reason)``
  - ``ConnectionCheckOutFailedEvent(ServerId, Reason)``
  - ``ConnectionCheckOutStartedEvent(ServerId)``
  - ``ConnectionReadyEvent(ConnectionId)``
  - ``ServerHeartbeatFailedEvent(ConnectionId, long, Throwable)``
  - ``ServerHeartbeatSucceededEvent(ConnectionId, BsonDocument, long)``

- Removes the ``errorLabels`` option from the ``WriteConcernError``
  class. This includes the ``addLabel()`` and ``getErrorLabels()`` methods and the
  constructor that includes an ``errorLabels`` parameter. Instead, you can use
  the error labels included in the ``MongoException`` object that contains the
  ``WriteConcernError``.

- Removes the following classes from the
  ``com.mongodb.event`` package:

  - ``ConnectionAddedEvent``
  - ``ConnectionPoolOpenedEvent``
  - ``ConnectionRemovedEvent``
  - ``ClusterListenerAdapter``
  - ``ConnectionPoolListenerAdapter``
  - ``ServerListenerAdapter``
  - ``ServerMonitorListenerAdapter``

  Because of these removals, the following methods were also removed from the
  ``ConnectionPoolListener`` interface:

  - ``connectionAdded``
  - ``connectionPoolOpened``
  - ``connectionRemoved``

>   For more information about the events package, see the `com.mongodb.event

Author: Reworded a little

>   package documentation <{+api+}/mongodb-driver-core/com/mongodb/event/package-summary.html>`__

Author: changed to "API documentation"

  that contains a ``List`` that receives a ``Chapter`` type parameter:

  .. code-block:: java

     public record Book(String title, List<Chapter> chapters) {}
     public record Chapter(Integer number, String text) {}

  Starting in this version, the codec deserializes data in the ``List`` into
  ``Chapter`` record classes instead of ``Document`` values.

.. _scala-breaking-changes-v4.7:

Version 4.7 Breaking Changes
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This driver version introduces the following breaking changes:

- The ``setWindowFields`` builder API is no longer beta. The new builder
  breaks binary and source compatibility. See the
  `Aggregates API documentation <https://mongodb.github.io/mongo-java-driver/4.7/apidocs/mongodb-driver-core/com/mongodb/client/model/Aggregates.html>`__
  for information the new ``setWindowFields()`` method signatures.

  If your application uses this builder in a version earlier than v4.7, update
  your source code to use the new method signature and rebuild your binary.

.. _scala-breaking-changes-v4.2:

Version 4.2 Breaking Changes
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This driver version introduces the following breaking changes:

> - The ``ObjectId`` class and its ``serialVersionUID`` field were updated

Reviewer: I: Use active voice

> - The ``ObjectId`` class and its ``serialVersionUID`` field were updated
>   to use a new format that minimizes serialization compatibility issues

Reviewer: ```suggestion
- The ``ObjectId`` class and its ``serialVersionUID`` field use a new format that minimizes serialization compatibility issues
```

  across different versions of the driver.

>   If an application using this version or later of the driver attempts to

Reviewer: S:

```suggestion
  If an application using driver version 4.2 or later of the driver attempts to
```

  perform Java Object Serialization on any objects that contain an
  ``ObjectId`` and were serialized by a prior version of the driver, Java
  throws an ``InvalidClassException``.

>   To learn more about Java Object Serialization, see the Java
>   Documentation on `Serializable Objects <https://docs.oracle.com/javase/tutorial/jndi/objects/serial.html>`__.

Reviewer: ```suggestion
  To learn more about Java Object Serialization, see `Serializable Objects <https://docs.oracle.com/javase/tutorial/jndi/objects/serial.html>`__ in the
  Java documentation.
```

.. _scala-breaking-changes-v4.0:

Version 4.0 Breaking Changes
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This driver version introduces the following breaking changes:

> - Several classes and methods marked as deprecated in the 3.12 release

Reviewer: I: use active voice throughout this section

  were removed in this version.
- The insert helper methods return an insert result object instead of
  ``void``.

> - The ``toJson()`` methods on ``BsonDocument``, ``Document``, and
>   ``DbObject`` return a relaxed JSON format instead of a strict JSON

Reviewer: ```suggestion
- The ``toJson()`` methods in the ``BsonDocument``, ``Document``, and
  ``DbObject`` classes return a relaxed JSON format instead of a strict JSON
```

  format. This makes the JSON documents more readable, but can make it more
  difficult to identify the BSON type information, such as the difference
  between a 32-bit and 64-bit integer. If your application relies on the
  strict JSON format, use the strict mode when reading or writing data.

> - The default BSON representation of ``java.util.UUID`` value was changed

Reviewer: ```suggestion
- The default BSON representation of ``java.util.UUID`` values was changed
```

  from ``JAVA_LEGACY`` to ``UNSPECIFIED``. Applications that store or retrieve
  UUID values must explicitly specify which representation to use. You can
  specify the representation in the ``uuidRepresentation`` property of

>   ``MongoClientSettings``.

Reviewer: ```suggestion
  a ``MongoClientSettings`` object.
```

  The UUID representation that you specify strictly controls how the driver
  decodes UUIDs. In previous versions of the driver, if you specified the
  ``JAVA_LEGACY`` representation, the driver would decode binary objects
  of subtypes 3 and 4 as UUIDs. In version 4.0, the ``JAVA_LEGACY``

>   representation works only with subtype 3.

Reviewer: S: Switch current and previous functionality

  For a list of members in the ``UuidRepresentation`` enum, see the

>   `v4.0 API documentation <https://mongodb.github.io/mongo-java-driver/4.0/apidocs/bson/org/bson/UuidRepresentation.html>`__.

Reviewer: ```suggestion
  `v4.0 API documentation. <https://mongodb.github.io/mongo-java-driver/4.0/apidocs/bson/org/bson/UuidRepresentation.html>`__
```

- The connection pool no longer restricts the number of wait queue threads

>   or asynchronous tasks that require a connection to MongoDB. The
>   application will throttle requests as necessary rather than depend on
>   the driver to throw a ``MongoWaitQueueFullException``.

Reviewer: ```suggestion
  or asynchronous tasks that require a connection to MongoDB. The
  application throttles requests as necessary rather than depending on
  the driver to throw a ``MongoWaitQueueFullException``.
```

> - The driver no longer logs using the ``java.util.logging`` (JUL) package

Reviewer: Q: Is "JUL" used elsewhere in the page? Can we just leave it out?

  and only supports the SLF4J logging framework.
- The embedded and Android drivers were removed. If your application
  relies on these drivers, you must continue to use a 3.x Java driver
  version.

> - The uber JARs, ``mongo-java-driver`` and ``mongodb-driver``, are no

Reviewer: Q: Are these the only two uber JARs for the Java driver?

- If not, leave out the commas (because they would just be two examples of uber JARs).
- If so, I'd qualify "uber JARs":

```suggestion
- The uber JARs for the Java driver, ``mongo-java-driver`` and ``mongodb-driver``, are no
```

>   longer published. If your application relies on one of these, you must
>   switch to either ``mongodb-driver-sync`` or ``mongodb-driver-legacy``
>   depending on which API the application uses. Make sure you remove the
>   uber JARs from your dependencies.

Reviewer: ```suggestion
  longer published. If your application relies on either [or both?] of these uber JARs, remove
  it [/them] as a dependency and use one of the following packages instead:
  
  - If your application uses the current API, add the ``mongodb-driver-sync`` package as a dependency.
  - If your application uses the legacy API, add the ``mongodb-driver-legacy`` package as a dependency. 
```

please correct as needed!

> - Updates to several classes introduced binary compatibility breaks, such

Reviewer: ```suggestion
- Several classes introduce binary compatibility breaks, such
```

  as the method signature change to the insert helper methods. Recompile
  any classes that link to the driver against this version or later to ensure
  that they continue to work.

.. _scala-server-release-changes:

Server Release Compatibility Changes
------------------------------------

A server release compatibility change is a modification
to the {+driver-short+} that discontinues support for a set of
{+mdb-server+} versions.

The driver discontinues support for a {+mdb-server+} version after it reaches
end-of-life (EOL).

To learn more about the MongoDB support for EOL products,
see the `Legacy Support Policy <https://www.mongodb.com/support-policy/legacy>`__.

.. _scala-server-8.1-incompatibility:

Server Version 8.1 Support Changes
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

You cannot use a 3.x version of the {+driver-short+} to connect to a

> deployment running {+mdb-server+} v8.1. Starting in {+mdb-server+} v8.1,

Reviewer: ```suggestion
MongoDB deployment running {+mdb-server+} v8.1. Starting in {+mdb-server+} v8.1,
```

  :ref:`Breaking Changes <scala-breaking-changes>` section. To learn
  more about {+mdb-server+} release compatibility changes, see the
  :ref:`<scala-server-release-changes>` section.

.. tip::

   To minimize the number of changes your application might require when
   you upgrade driver versions in the future, use the
   :ref:`{+stable-api+}. <scala-stable-api>`

.. _scala-breaking-changes:

Breaking Changes
----------------

A breaking change is a modification in a convention or behavior in
a specific version of the driver that might prevent your application from
working properly if not addressed before upgrading.

The breaking changes in this section are categorized by the driver version that
introduced them. When upgrading driver versions, address all the breaking
changes between the current and upgrade versions. For example, if you
are upgrading the driver from v4.0 to v4.7, address all breaking changes from
the version after v4.0 including any listed for v4.7.

.. _scala-breaking-changes-v5.2:

Version 5.2 Breaking Changes
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

This driver version introduces the following breaking changes:

> - Revises the `mongodb-crypt <https://mvnrepository.com/artifact/org.mongodb/mongodb-crypt>`__

Reviewer 2: We also removed support for MongoDB 3.6.

  - ``ServerListenerAdapter``
  - ``ServerMonitorListenerAdapter``

  The driver also removes the following related methods from the
  ``ConnectionPoolListener`` interface:

  - ``connectionAdded()``
  - ``connectionPoolOpened()``
  - ``connectionRemoved()``

  For more information about the ``com.mongodb.event`` package, see the

>   `API documentation <{+core-api+}/com/mongodb/event/package-summary.html>`__.

Reviewer: ```suggestion
  `API documentation. <{+core-api+}/com/mongodb/event/package-summary.html>`__
```

.. _scala-breaking-changes-v5.0-list-collections:

- Adds the ``authorizedCollection`` option for the ``listCollections`` command.
  This introduces a breaking binary change in the ``MongoDatabase.listCollectionNames()``

>   methods. This change does not require any changes to source code, but you must

Reviewer: Q: Should this be "method", or is there more than one? Also applies to the next line.

Author: Looks like just one

This driver version introduces the following breaking changes:

- Removes several classes and methods marked as deprecated in the 3.12 release.

> - Modifies insert helper methods to return an insert result object instead of

Reviewer: S: Might be easier to parse if you used the actual class name: "to return an ``InsertResult`` object" (or whatever it's called)
